tutorials/synthetic-retrieval-evaluation-customization/filters.py

```diff
@@ -0,0 +1,197 @@
+            elif isinstance(text, str):
+                embeddings = response.data[0].embedding
+            return embeddings
+        else:
+            return []
+
+    def _calc_similarity_nim(self, context, question):
+        # cosine similarity
+        doc_embed = self._get_nim_embedding(text=context, input_type="passage")
+        q_embed = self._get_nim_embedding(text=question, input_type="query")
+        if isinstance(context, list) and isinstance(question, list):
+            if doc_embed and q_embed:
+                sim = np.diag(np.dot(np.array(doc_embed), np.array(q_embed).T))
+            else:
+                sim = np.zeros(len(context))
+        else:
+            if doc_embed and q_embed:
+                sim = np.dot(doc_embed, q_embed)
+            else:
+                sim = 0.0
+
+        return sim
+
+
+# ----------------------------------------------------------------------------80
+# ----------------------- Answerability Filter ---------------------------------
+# ----------------------------------------------------------------------------80
+
+
+class AnswerabilityFilter(DocumentFilter):
+
+    def __init__(self, cfg: DictConfig, text_fields: List[str] = ["text", "question"]):
+
+        self._name = "answerability_filter"
+        if "answerability_filter" in cfg:
+            self.filter_cfg = cfg["answerability_filter"]["filter_cfg"]
+        else:
+            raise Exception("Error: Config doesn't have answerability filter")
+        self.base_url = self.filter_cfg.base_url
+        self.api_key = self.filter_cfg.api_key
+        self.model_name = self.filter_cfg.model_name
+        self.system_prompt = self.filter_cfg.system_prompt
+        self.user_prompt_template = self.filter_cfg.user_prompt_template
+        self.num_criteria = self.filter_cfg.num_criteria
+
+        self.client = OpenAI(base_url=self.base_url, api_key=self.api_key)
+        self.text_fields = text_fields
+
+    @batched
+    def score_document(self, df: Union[pd.DataFrame, dd.DataFrame]):
+        return df.apply(
+            lambda row: self._llm_as_judge(
+                row[self.text_fields[0]], row[self.text_fields[1]]
+            ),
+            axis=1,
+        )
+
+    @batched
+    def keep_document(self, scores: pd.Series):
+
+        def _keep_document(score: str):
+            is_keep = True  # default is to keep
+            try:
+                json_ans = json.loads(scores)
+                for i in range(self.num_criteria):
+                    if json_ans[f"criterion_{i+1}"] != "Y":
+                        # filter out data if any of the criteria fails
+                        is_keep = False  # filter out
+                        break
+            except Exception as e:
+                pass # TODO log the errors
+                # print(f"Parse error {e}")
+                # if there is a parse error, keep the document
+
+            return is_keep
+
+        return scores.apply(_keep_document)
```

Comment:
Is there a reason you don't remove the `@batched` decorator and make `_keep_document` `keep_document`?